config/v1/types_authentication.go

```diff
@@ -565,27 +601,44 @@ type PrefixedClaimMapping struct {
 	Prefix string `json:"prefix"`
 }
 
+// TokenValidationRuleType defines the type of token validation rule.
+//
+// +kubebuilder:validation:Enum=RequiredClaim;Expression
 type TokenValidationRuleType string
 
 const (
-	TokenValidationRuleTypeRequiredClaim = "RequiredClaim"
+	TokenValidationRuleRequiredClaim TokenValidationRuleType = "RequiredClaim"
+	TokenValidationRuleExpression    TokenValidationRuleType = "Expression"
 )
 
+// TokenClaimValidationRule represents a validation rule based on token claims.
+// If type is RequiredClaim, requiredClaim must be set.
+// If type is Expression, expressionRule must be set.
+//
+// +kubebuilder:validation:XValidation:rule="self.type != 'RequiredClaim' || has(self.requiredClaim)",message="requiredClaim must be set when type is 'RequiredClaim'"
+// +kubebuilder:validation:XValidation:rule="self.type != 'Expression' || has(self.expressionRule)",message="expressionRule must be set when type is 'Expression'"
+
 type TokenClaimValidationRule struct {
 	// type sets the type of the validation rule
 	//
 	// +kubebuilder:validation:Enum={"RequiredClaim"}
```

Comment:
You can't have two enum markers on a field and on a type as it will result in unexpected enum constraints. Pick a single location where the allowed values should be specified.

If you are updating the allowed values of this field, you should also update the GoDoc.